release-notes.txt: add 2020.07 release notes (#14535)
RIOT-OS/RIOT: release-notes.txt

RIOT-2020.07 - Release Notes
============================
RIOT is a multi-threading operating system which enables soft real-time
capabilities and comes with support for a range of devices that are typically
found in the Internet of Things: 8-bit and 16-bit microcontrollers as well as
light-weight 32-bit processors.

RIOT is based on the following design principles: energy-efficiency, soft
real-time capabilities, small memory footprint, modularity, and uniform API
access, independent of the underlying hardware (with partial POSIX compliance).

RIOT is developed by an international open-source community which is
independent of specific vendors (e.g. similarly to the Linux community) and is
licensed with a non-viral copyleft license (LGPLv2.1), which allows indirect
business models around the free open-source software platform provided by RIOT.

About this release:
===================

> The 2020.07 release includes put a focus on the improvement of automated

Reviewer: ```suggestion
The 2020.07 release puts a focus on the improvement of automated
```
?

testing. In that vain various new CI integrations such as CircleCI for
documentation building and online presentation and Github Actions to check the
tooling of RIOT where merged. The `riotctrl_shell` python module allows to
abstract shell commands as python methods for testing using the newly created
`riotctrl` python package [1]. Kconfig migration reached phase 2 with various
board features being exposed to Kconfig. New network protocols were ported for
RIOT such as MQTT—in form of the `paho-mqtt` package and the lookup client

> component for CoRE RD `cord_lc`. The OpenWSN network stack with 6TSCH support

Reviewer: ```suggestion
component for CoRE RD `cord_lc`. The OpenWSN network stack with 6TiSCH support
```

> was reintegrated into RIOT. Suppotr for several new boards and new sensors was

Reviewer: ```suggestion
was reintegrated into RIOT. Support for several new boards and new sensors was
```

+ serpente: add support
+ sltb001a: enable CCS811 sensor
+ sodaq-sff: add support for SODAQ SFF
+ stm32f429i-disco: add support
+ weact-f411ce: add support
+ wemos-zero: add support

CPU
---
* arm7_common: Update to inlineable IRQ API
* atmega_common: Update to inlineable IRQ API
* cortexm:
  + Allow to define reserved fields in CortexM vector table
  * "free" SVC
* cc26x2_cc13x2:
  * add oscillator switching functions
  * add periph_i2c implementation
  * trim device registers on `cpu_init`
* esp32:
  * support of multiple heaps for newlib malloc
  * replace RTC implementation by RTT implementation
* fe310: migrate to inlined IRQ API
+ kinetis: implement power modes for pm_layered
* lpc2387:
  * PM: enable SLEEP & POWERDOWN mode
  * timer: use lpc2387_pclk_scale()
+ mips_pic32_common: Implement GPIO IRQ
+ native:
  + add periph_eeprom driver implementation
  * rename trace -> backtrace
+ nrf52:
  + add black magic probe config
  + add missing vendor files
  + implement EasyDMA-based SPI peripheral implemenation
+ nrf52811: initial support
* nrf5x_common:
  * add configurable RTT_FREQUENCY
  * enable bias correction in hwrng
* sam0_common:
  + add DMA peripheral driver

>   + add support for saml21, samd5x

Reviewer 2: ```suggestion
  + add PWM support for saml21, samd5x
```

  + implement periph/dac
  * Make SPI peripheral DMA compatible

>   * move PWM to common code

Reviewer 2: ```suggestion
```
the location of `pwm.c` isn't that noteworthy

  * spi: move clk pin muxing into spi_acquire / spi_release
  + UART: implement inverted RX & TX
* samd21: PWM don't hard-code number of channels to 3
* samd5x: allow to use XOSC as clock source
* saml1x: pm: set deep flag
+ stm32:
  + add QDEC support for CPU_FAM_STM32F1
  + add non-blocking UART
  + implement reset to bootloader
  * filter availability of RNG by CPU not by board
  * stm32_common: make RTT_FREQUENCY configurable
  + stm32f4: Add DMA config for SPI

>   + stm32g4: add support

Reviewer 3: I'm missing something along the lines of
```
  + Decrease latency in SPI DMA path
```

Something that describes the effect of both https://github.com/RIOT-OS/RIOT/pull/14096 and https://github.com/RIOT-OS/RIOT/pull/14097 in a single note. Feel free to reword my suggestion to something that makes sense for the reader

Author: See my latest change.

  + Add fuzzing application for gcoap
+ ci: setup github actions
+ sys/test_utils/dummy_thread: initial commit
+ tests/irq_disable_restore: Add test for irq_disable() / irq_restore()
+ tests/periph_spi: add benchmark for acquire / release
+ tests/periph_spi_dma: Add test for SPI with DMA
+ uncrustify: add auto uncrustify with blacklist

Kconfig migration
-----------------
+ armv7_m/cortexm: declare CPU_ARCH and CPU_CORE
+ boards: Model features in Kconfig
  + arduino-leonardo, arduino-mega2560, atmega128x-based, atmega256rfr2-based,
    atmega328p-based, cc2538-based, cc26xx_cc13xx-based, efm32-based, esp-based,
    fe310-based, kinetis-based, msp430-based, samr21-xpro, slwstk6220a
* kconfig: Declare common feature symbols
+ drivers: Expose Configurations to Kconfig
  + ads101x, at, at86rf215, bmx055, cc110x, dose.h, fxos8700, hdc1000, isl29020,
    l3g4200d, lpsxxx, mag3110, mma8x5x, motor_driver, mtd_sdcard, opt3001,
    pn532, rn2xx3, slipdev, sps30, tcs37727, tmp00x
+ networking: Expose configurations to Kconfig
  + asymcute, coap, csma
  + gnrc:
    + gomach, lwmac, mac, pktdump, pktbuf, rpl, tcp
+ usbus/cdc/ecm: Expose configurations to Kconfig

API Changes
===========

> - cpu/sam0_common: drop prescaler from timer config

Reviewer 2: ```suggestion
- cpu/sam0_common: drop prescaler from timer config
- cpu/sam0_common: PWM config now requires GCLK source 
```

Author: Which PR was that?

Author: Also https://github.com/RIOT-OS/RIOT/pull/14007? Then the label for API change was missing ;-)

Author: Clarified offline. It was indeed part of #14007

RIOT-2020.07 - Release Notes
============================
RIOT is a multi-threading operating system which enables soft real-time
capabilities and comes with support for a range of devices that are typically
found in the Internet of Things: 8-bit and 16-bit microcontrollers as well as
light-weight 32-bit processors.

RIOT is based on the following design principles: energy-efficiency, soft
real-time capabilities, small memory footprint, modularity, and uniform API
access, independent of the underlying hardware (with partial POSIX compliance).

RIOT is developed by an international open-source community which is
independent of specific vendors (e.g. similarly to the Linux community) and is
licensed with a non-viral copyleft license (LGPLv2.1), which allows indirect
business models around the free open-source software platform provided by RIOT.

About this release:
===================

The 2020.07 release puts a focus on the improvement of automated testing. In
that vain various new CI integrations such as CircleCI for documentation
building and online presentation and Github Actions to check the tooling of

> RIOT where merged. The `riotctrl_shell` python module allows to abstract shell

Reviewer 4: ```suggestion
RIOT were merged. The `riotctrl_shell` python module allows to abstract shell
```

  * stm32_common: make RTT_FREQUENCY configurable
  + stm32f4: Add DMA config for SPI
  + stm32g4: add support

Device Drivers
--------------
+ atwinc15x0: add WiFi netdev driver
+ at: Add at_urc_isr module to process URCs upon arrival
+ at24cxxx: MTD wrapper for at24cxxx
+ at86rf215:
  + implement MR-O-QPSK
  + implement MR-OFDM
+ at86rf2xx: add support for high data rates
+ dfplayer: New driver for the DFPlayer Mini MP3 player
* ethos: use luid_get_eui48()
+ edbg_eui: add driver to get MAC address from Atmel EDBG
+ mhz19: initial support
+ mrf24j40: add Turbo Mode
+ mtd_spi_eeprom: add MTD wrapper for AT25XXX EEPROMs
+ pca9633: add new driver
+ periph/timer: add timer_set_periodic()
+ periph/i2c: add periph_i2c_reconfigure feature & implementation for sam0
+ periph/rtt: introduce RTT_MIN_OFFSET
+ periph/spi: add periph_spi_reconfigure feature & implementation for sam0
* pulse_counter: Use C11 atomics & bugfix
+ sdp3x : Add driver for sdp3x sensor

Build System / Tooling
----------------------
+ make:
  + add code generator targets
  + add JLink support for stm32
  * docker.inc.mk: add and use generic docker_run_make
  + allow multiple external board dirs

>   + info-global.inc.mk: add info-boards-features-blacklisted

Reviewer 4: Do you think we should include #14351?

Author: Done

RIOT-2020.07 - Release Notes
============================
RIOT is a multi-threading operating system which enables soft real-time
capabilities and comes with support for a range of devices that are typically
found in the Internet of Things: 8-bit and 16-bit microcontrollers as well as
light-weight 32-bit processors.

RIOT is based on the following design principles: energy-efficiency, soft
real-time capabilities, small memory footprint, modularity, and uniform API
access, independent of the underlying hardware (with partial POSIX compliance).

RIOT is developed by an international open-source community which is
independent of specific vendors (e.g. similarly to the Linux community) and is
licensed with a non-viral copyleft license (LGPLv2.1), which allows indirect
business models around the free open-source software platform provided by RIOT.

About this release:
===================

The 2020.07 release puts a focus on the improvement of automated testing. In

> that vain various new CI integrations such as CircleCI for documentation

Reviewer 5: ```suggestion
that vein, various new CI integrations such as CircleCI for documentation
```

> building and online presentation and Github Actions to check the tooling of

Reviewer 5: ```suggestion
building and online presentation, and Github Actions to check the tooling of
```
(eases the reading flow given there's already more strongly binding "and")

> RIOT were merged. The `riotctrl_shell` python module allows to abstract shell

Reviewer: ```suggestion
RIOT when merged. The `riotctrl_shell` python module allows to abstract shell
```

Author: Addressed and squashed directly.

Reviewer: Now that I'm reading this again, the initial word `were` was ok.

commands as python methods for testing using the newly created `riotctrl`
python package [1]. Kconfig migration reached phase 2 with various board
features being exposed to Kconfig. New network protocols were ported for RIOT

> such as MQTT—in form of the `paho-mqtt` package and the lookup client

Reviewer 5: ```suggestion
such as MQTT (in form of the `paho-mqtt` package) and the lookup client
```

Kconfig migration
-----------------
+ armv7_m/cortexm: declare CPU_ARCH and CPU_CORE
+ boards: Model features in Kconfig
  + arduino-leonardo, arduino-mega2560, atmega128x-based, atmega256rfr2-based,
    atmega328p-based, cc2538-based, cc26xx_cc13xx-based, efm32-based, esp-based,
    fe310-based, kinetis-based, msp430-based, samr21-xpro, slwstk6220a
* kconfig: Declare common feature symbols
+ drivers: Expose Configurations to Kconfig
  + ads101x, at, at86rf215, bmx055, cc110x, dose.h, fxos8700, hdc1000, isl29020,
    l3g4200d, lpsxxx, mag3110, mma8x5x, motor_driver, mtd_sdcard, opt3001,
    pn532, rn2xx3, slipdev, sps30, tcs37727, tmp00x
+ networking: Expose configurations to Kconfig
  + asymcute, coap, csma
  + gnrc:
    + gomach, lwmac, mac, pktdump, pktbuf, rpl, tcp
+ usbus/cdc/ecm: Expose configurations to Kconfig

API Changes
===========
- cpu/sam0_common: drop prescaler from timer config
- cpu/sam0_common: PWM config now requires GCLK source
- drivers/mtd: Change API to return 0 on success
- drivers/mtd_mapper: count offset in sectors
- net/sock_dtls: provide getter function for UDP sock
- net/sock_dtls: add timeout to sock_dtls_send and add sock_dtls_session_init

> 

Reviewer 5: Does this large diff belong here? It's way more detailed than all the rest, and looks a bit like a copy-pasting slip.

Author: The rest are either simple changes or additions. This change is a larger API overhaul and thus justifies a more detailed listing of the changes.

Author: (see also larger API changes in other releases below)
